pkg/cli/nodes_connectivity.go

```diff
@@ -0,0 +1,403 @@
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/kustomize/api/types"
+
+	"github.com/google/uuid"
+	"github.com/replicatedhq/plumber/v2"
+	"github.com/spf13/cobra"
+
+	"github.com/replicatedhq/kurl/pkg/k8sutil"
+	"github.com/replicatedhq/kurl/pkg/static/nodes_connectivity"
+)
+
+var usageExamples = `
+# Test if all nodes can reach all other nodes using tcp in port 6472.
+kurl netutil nodes-connectivity --port 6472 --proto tcp
+# Test if all nodes can reach all other nodes using udp in port 7788.
+kurl netutil nodes-connectivity --port 7788 --proto udp
+`
+
+const (
+	listenersSelector = "name=nodes-connectivity-listener"
+	pingerSelector    = "name=nodes-connectivity-pinger"
+)
+
+type logLine struct {
+	message string
+	err     error
+}
+
+type nodeConnectivityOptions struct {
+	namespace string
+	proto     string
+	attempts  int
+	cliset    kubernetes.Interface
+	image     string
+	port      int32
+	cli       client.Client
+	wait      time.Duration
+}
+
+func newNetutilNodesConnectivity(_ CLI) *cobra.Command {
+	var opts nodeConnectivityOptions
+	cmd := &cobra.Command{
+		Use:     "nodes-connectivity",
+		Short:   "Tests if all nodes can reach all other nodes using the provided protocol in the provided port",
+		Example: usageExamples,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), syscall.SIGTERM, syscall.SIGINT)
+			defer cancel()
+
+			if opts.port == 0 {
+				return fmt.Errorf("--port flag is required.")
+			}
+			opts.proto = strings.ToUpper(opts.proto)
+			proto := corev1.Protocol(opts.proto)
+			if proto != corev1.ProtocolTCP && proto != corev1.ProtocolUDP {
+				return fmt.Errorf("--protocol must be either tcp or udp.")
+			}
+			opts.wait = time.Second
+			if corev1.Protocol(opts.proto) == corev1.ProtocolUDP {
+				opts.wait = 5 * time.Second
+			}
+
+			cli, err := client.New(config.GetConfigOrDie(), client.Options{})
+			if err != nil {
+				return fmt.Errorf("failed to create kubernetes client: %w", err)
+			}
+			opts.cli = cli
+
+			// XXX controller-runtime kubernetes client does not have access to subresources that span beneath
+			// crud objects (e.g. pod logs). we have to create a kubernetes clientset, and that is why we can't
+			// have nice things.
+			cliset, err := kubernetes.NewForConfig(config.GetConfigOrDie())
+			if err != nil {
+				log.Printf("Failed to create kubernetes client set: %s", err)
+				os.Exit(1)
+			}
+			opts.cliset = cliset
+
+			cleanup := func() {
+				ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+				defer cancel()
+				if err := deleteListeners(ctx, opts); err != nil {
+					log.Printf("Failed to delete DaemonSet listeners overlay: %s", err)
```

Comment:
most of my comments in this pr are around log statements. i am only going to make one comment here and can you adjust the other lines accordingly?

i think its really time we standardize on how we log in a cli. i think id like to use the functions build into cobra.Command (cmd.Print) unless you have objections. Is it time to make a type for reuse in all commands to make it more explicit?

Reply:
I agree with you. I think we can address this need for a standardised logger type in a separated issue, what do you think ? To solve your initial concern I have changed this PR to use `cmd.Printf` everywhere. To be honest I don't like it because it does not include timestamps but I am OK in having it here, once we have our own type we can discuss these details.